### DISCUSSION TITLE
[Bug] Not enough memory / export of notes

### DISCUSSION THREAD

**Discussion by sjgknight** [windingwind/zotero-better-notes#437] [Q&A]
**Describe the bug**
On running a template I'm getting a 'not enough memory resources are available to process this command' error, with a failed to encode base64 data (this is a template with annotation images). 

**To Reproduce**
1. Run template [Item] Annotations by tag
2. The template works but at some point it just crashed and zotero needed to be restarted 

I tried running the template in blocks of ~50 items a time, and this definitely helped, but I guess the first blocks are still live links so perhaps that has an impact on memory (I'll try unlinking them maybe, or run the chunks in multiple separate notes).

**Desktop (please complete the following information):**
 - OS: Windows

**Additional context**
I wonder if it would be possible to do something like 'send template to export', to run a template that writes an external file?  
Alternatively, whether the 'export' function could be modified in some ways to make use of some of the benefits of better notes, e.g., at the moment they don't include tags or parent item metadata, but these could be included (ideally as yaml headers), and then users can work with an external note taking tool.

**Comment by windingwind** [comment]
When you embed an image to a note, Zotero will encode it to base64 to display. I’m not pretty sure how this works but it is possible to happen when there are too many images in a note. Besides, Zotero has a note length limit for syncing, we’d better not make them too long.

About export a template result directly, I am refactor the code for two path syncing and we will use a different backend to process md/note conversions in the future. We could expect it to be possible.

**Comment by windingwind** [comment]
Besides, the more I dive deeper into the Zotero note in richtext, the more shortcomings and limitations I find with it. Maybe in a further future, we could migrate to a pure md note system.

**Comment by sjgknight** [comment]
Thanks. This is a slightly separate question but it's effectively a workaround for this issue so I hope ok to tag it on here. Specifically, it's fantastic I can export the notes+images, but I would ideally also export the tags and parent itemkey (in an ideal world as a yaml header, but I can reformat for that myself). I can't see the template in the template editor.

Is it possible to edit the export-template in better notes?  

If not, a slightly kludgey approach, in the `ExportMDFileName` template, could we do something like `${noteItem.key}${noteItem.getNoteTags}${noteItem.citekey}.md`  (fields for illustration), that way I can use the filename to link the noteItem to the parentItem by key, and extract the tags.  The tags are returned as [Object object], and I can't workout what the field for parent.key is?

**Comment by windingwind** [comment]
With the new export backend, the tags are automatically added to the YAML header. The memory bug (haven't tested it, suppose it is a bug) may not be solved as it's not a common use case and is hard to track.

If no updates, I'll close this issue. If you think this thread would be helpful to others, converting it to a discussion would be better.

Thanks for the feedback!

**Comment by sjgknight** [comment]
@windingwind thanks! The YAML header on the notes is fantastic, and seems a great workaround.  I think this is likely to be useful to others, so may be good to convert to a discussion (or/and, to link to the closed issue from the template that compiles from tags). 

Thanks again
